File: solutions/problem_1593_split_a_string_into_the_max_number_of_unique_substrings/solution.py
"""
Problem 1593: Split a String Into the Max Number of Unique Substrings
===================================================================
Difficulty: Medium
Tags: Hash Table, String, Backtracking
Pattern: Backtracking

Time Complexity:  O(k^n) or O(n!)
Space Complexity: O(n)
"""

class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        # Backtracking - O(2^n) or O(n!) time
        result = []
        
        def backtrack(path, start):
            result.append(path[:])
            for i in range(start, len(s)):
                path.append(s[i])
                backtrack(path, i + 1)
                path.pop()
        
        backtrack([], 0)
        return result

File: solutions/problem_1593_split_a_string_into_the_max_number_of_unique_substrings/solution.cpp
/*
 * Problem 1593: Split a String Into the Max Number of Unique Substrings
 * ===================================================================
 * Difficulty: Medium
 * Tags: Hash Table, String, Backtracking
 * Pattern: Backtracking
 *
 * Time Complexity:  O(k^n) or O(n!)
 * Space Complexity: O(n)
 */

#include <functional>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int maxUniqueSplit(string& s) {
        // Backtracking - O(2^n) or O(n!) time
        vector<vector<int>> result;
        vector<int> path;
        function<void(int)> backtrack = [&](int start) {
            result.push_back(path);
            for (int i = start; i < (int)s.size(); i++) {
                path.push_back(s[i]);
                backtrack(i + 1);
                path.pop_back();
            }
        };
        backtrack(0);
        return result;
    }
};
